src/io/flutter/run/LauncherState.java

```diff
@@ -238,33 +254,71 @@ protected boolean canRun(C config) {
     @Override
     protected final RunContentDescriptor doExecute(@NotNull RunProfileState state, @NotNull ExecutionEnvironment env)
       throws ExecutionException {
-      if (!(state instanceof Launcher)) {
+      if (!(state instanceof LauncherState)) {
         LOG.error("unexpected RunProfileState: " + state.getClass());
         return null;
       }
-      final Launcher launcher = (Launcher)state;
-      return launcher.launch(env);
+
+      final LauncherState launcherState = (LauncherState)state;
+      final String executorId = env.getExecutor().getId();
+
+      // See if we should issue a hot-reload.
+      final List<RunContentDescriptor> runningProcesses =
+        ExecutionManager.getInstance(env.getProject()).getContentManager().getAllDescriptors();
+
+      for (RunContentDescriptor descriptor : runningProcesses) {
+        final ProcessHandler process = descriptor.getProcessHandler();
+        if (process != null && !process.isProcessTerminated() && process.getUserData(FLUTTER_RUN_CONFIG_KEY) == launcherState.runConfig) {
+          final String launchMode = process.getUserData(FLUTTER_LAUNCH_MODE_KEY);
+          final FlutterApp app = process.getUserData(FLUTTER_APP_KEY);
+          if (executorId.equals(launchMode) && app != null && app.getMode().isReloadEnabled()) {
+            if (!app.isStarted()) {
+              return null;
+            }
+
+            // Map a re-run action to a flutter full restart.
+            FileDocumentManager.getInstance().saveAllDocuments();
+            app.performRestartApp();
+
+            return null;
+          }
+        }
+      }
+
+      // Else, launch the app.
+      return launcherState.launch(env);
     }
 
     /**
      * Returns true if any processes are running that were launched from the given RunConfig.
      */
-    private static boolean isRunning(RunConfig config) {
+    @SuppressWarnings("SameParameterValue")
+    private static boolean isRunning(RunConfig config, String executorId, boolean ignoreSameLaunchMode) {
```

Comment:
The ignoreSameLaunchMode parameter here is confusing. How about a more general-purpose method instead?

```
/**
 * Returns the currently running app for the given RunConfig, if any.
 */
@Nullable
private static FlutterApp getRunningApp(RunConfig config);
```

Then the caller can decide what to do based on the running FlutterApp.